### PR TITLE
Collection expressions: use inline array for span argument when the argument cannot escape

### PR DESCRIPTION
**Pull request by cston**
*Split out from #69291.*

Relates to test plan https://github.com/dotnet/roslyn/issues/66418

### CODE REVIEW

**Comment by Sergio0694** [comment]
When using this on a runtime without `[InlineArray]`, would this show an error or silently allocate an array? 🤔

**Comment by cston** [comment]
> When using this on a runtime without `[InlineArray]`, would this show an error or silently allocate an array?

The proposal implemented here is to allocate an array on heap when the runtime does not support inline arrays.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs [hunk 1]
```diff
@@ -218,10 +217,19 @@ private BoundExpression VisitCollectionBuilderCollectionExpression(BoundCollecti
                 call.Type);
         }
 
-        private BoundExpression CreateAndPopulateInlineArray(
+        private bool ShouldUseInlineArray(BoundCollectionExpression node)
+        {
+            var elements = node.Elements;
+            return elements.Length > 0 &&
+                !elements.Any(i => i is BoundCollectionExpressionSpreadElement) &&
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:224]: I'm not sure what you're asking here. 


"but didn't this ended up being an expression?"

What are you referring to here?

**Comment by cston** [comment]
Replaced by https://github.com/dotnet/roslyn/pull/69412.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alrz) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:224]: Sorry, I was asking about the syntax model for collection expressions and whether or not CollectionElementSyntax needs to exist to distinguish between spread and expression elements. Would it make sense to consider a SpreadExpression here? As mentioned above I don't think this involves any additional complexity in the parser.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:224]: I'm not seeing the value in a SpreadExpression.  We'd also then need a KeyValueExpression.  Neither of these could show up anywhere else, so i'm not sure why that would be a good thing.  Conversely, having specific element types for a list matches what we do elsewhere, and provides clear informatino as to the strong types, and expected shapes, one would expect when processing these lists.

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:224]: NOte: further things might be a construct like `if (x) y` allowed as a collection-expression element.  Making expressions out of all of these things (even though they would only be viable in one parent) is not a good thing IMO.

**Pull request closed without merging**